File: 2025-06-June-GFG-POTD/June 26 - Game with String/Solution.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <string>
using namespace std;

class Solution {
public:
    int minValue(string s, int k) {
        vector<int> freq(26, 0);

        for (char c : s) {
            freq[c - 'a']++;
        }

        priority_queue<int> maxHeap;
        for (int f : freq) {
            if (f > 0) {
                maxHeap.push(f);
            }
        }

        while (k > 0 && !maxHeap.empty()) {
            int top = maxHeap.top();
            maxHeap.pop();
            top--;
            k--;
            if (top > 0) {
                maxHeap.push(top);
            }
        }

        int result = 0;
        while (!maxHeap.empty()) {
            int f = maxHeap.top();
            maxHeap.pop();
            result += f * f;
        }

        return result;
    }
};

File: 2025-06-June-GFG-POTD/June 26 - Game with String/Solution.java
class Solution {
    public int minValue(String s, int k) {
        // Frequency array for 26 lowercase characters
        int[] freq = new int[26];
        for (char c : s.toCharArray()) {
            freq[c - 'a']++;
        }

        // Max heap to store frequencies
        PriorityQueue<Integer> maxHeap = new PriorityQueue<>(Collections.reverseOrder());
        for (int f : freq) {
            if (f > 0) {
                maxHeap.add(f);
            }
        }

        // Remove k characters
        while (k > 0 && !maxHeap.isEmpty()) {
            int top = maxHeap.poll();
            top--; // Remove one character
            k--;
            if (top > 0) {
                maxHeap.add(top);
            }
        }

        // Calculate the sum of squares of remaining frequencies
        int result = 0;
        while (!maxHeap.isEmpty()) {
            int remaining = maxHeap.poll();
            result += remaining * remaining;
        }

        return result;
    }
}
